File: src/core/src/todo.ts
import { TagProps } from 'element-plus'
import { Avatars } from './avatar'
import { Option } from './common'
import { Annex } from './annex'
import { User } from './user'

/** 优先级等级枚举，用于定义任务或项目的紧急程度 */
export enum Priorities {
  Emergent = 'Emergent', // 紧急
  High = 'High',         // 高
  Mid = 'Mid',           // 中
  Low = 'Low'            // 低
}

/** 优先级选项数组，用于前端选择组件 */
export const priorityOptions = [
  {
    value: Priorities.Low,    // 低优先级
    label: Priorities.Low     // 显示标签为"Low"
  },
  {
    value: Priorities.Mid,    // 中优先级
    label: Priorities.Mid     // 显示标签为"Mid"
  },
  {
    value: Priorities.High,   // 高优先级
    label: Priorities.High    // 显示标签为"High"
  },
  {
    value: Priorities.Emergent, // 紧急优先级
    label: Priorities.Emergent  // 显示标签为"Emergent"
  }
]

/** 优先级类型定义，包含颜色和名称 */
export type Priority = {
  color: string         // 优先级颜色
  name: Priorities      // 优先级名称
}

/** 优先级的颜色映射器，用于根据优先级获取对应的颜色 */
const PriorityColorMap = new Map<Priorities, string>([
  [Priorities.Emergent, '#E86D5E'], // 紧急
  [Priorities.High, '#F69D50'],     // 高
  [Priorities.Mid, '#6CB6FF'],      // 中
  [Priorities.Low, '#ADAC9A']       // 低
])

/** 获取优先级颜色的函数 */
export const usePriorityColor = (priority: Priorities): string => {
  return PriorityColorMap.get(priority) || '#ADAC9A'
}

/** 日期类型定义，用于描述任务或事件的开始和结束时间 */
export type IDate = {
  start: string  // 开始时间
  end: string    // 结束时间
  during: number // 持续时间，以某种计量单位表示
}

/** 待办状态枚举，描述任务的当前状态 */
export enum Status {
  NOT_START = 'not start',   // 未开始
  IN_PROGRESS = 'in progress', // 进行中
  COMPLETED = 'completed',   // 已完成
  PENDING = 'pending',       // 阻塞中
  FAILED = 'failed'          // 失败
}

/** 状态和颜色的映射，用于根据任务状态获取对应的颜色 */
const StatusTypeMap = new Map<Status, string>([
  [Status.NOT_START, '#ADAC9A'],     // 未开始
  [Status.IN_PROGRESS, '#56D4DD'],   // 进行中
  [Status.COMPLETED, '#8DDB80'],     // 已完成
  [Status.PENDING, '#8EBAC7'],       // 阻塞中
  [Status.FAILED, '#FF5555']         // 失败
])

/** 获取任务状态颜色的函数 */
export const useStatus = (status: Status): string => {
  return StatusTypeMap.get(status) || '#ADAC9A'
}

/** 标签属性类型，用于定义任务或项目标签的视觉表现 */
export type ITagProps = {
  type: 'info' | 'success' | 'warning' | '' | 'danger' // 标签类型
  effect: 'dark' | 'light' | 'plain'                    // 标签效果
  label: string                                         // 标签文本
}

/** 标签效果选项数组，用于前端选择组件 */
export const effectOptions = [
  // 定义不同的标签效果选项
  { label: 'dark', value: 'dark' },
  { label: 'light', value: 'light' },
  { label: 'plain', value: 'plain' }
]

/** 标签类型选项数组，用于前端选择组件 */
export const typeOptions = [
  // 定义不同的标签类型选项
  { label: 'default', value: '' },
  { label: 'info', value: 'info' },
  { label: 'success', value: 'success' },
  { label: 'warning', value: 'warning' },
  { label: 'danger', value: 'danger' }
]

/** 待办事项类型定义，描述一个待办事项的各种属性 */
export type Todo = {
  id?: string                      // 待办事项的唯一标识
  owner: string                    // 待办事项的所有者
  name: string                     // 待办事项的名称
  priority: Priorities             // 待办事项的优先级
  reviewers: Array<User>           // 审核人列表
  performers: Array<User>          // 执行人列表
  date: IDate                      // 待办事项的日期
  tags: Array<ITagProps>           // 待办事项的标签列表
  status: Status                   // 待办事项的状态
  description: Option<string>      // 待办事项的描述
  information: Option<string>      // 待办事项的额外信息
  annexs: Option<Array<Annex>>     // 附件列表
  isFocus: boolean                 // 是否为关注的待办事项
}

/** 待办事项箱类型定义，用于分类存放不同优先级和状态的待办事项 */
export type TodoBox = {
  low: Array<Todo>     // 低优先级的待办事项列表
  mid: Array<Todo>     // 中等优先级的待办事项列表
  fatal: Array<Todo>   // 高优先级和紧急优先级的待办事项列表
  focus: Array<Todo>   // 关注的待办事项列表
  history: Array<Todo> // 历史待办事项列表
}

File: back_end/todo/src/lib/entry/po/todo.rs
// 引入当前crate中其他模块或结构体的路径
use crate::lib::entry::dto;

// 引入同级或父级目录中的模块或结构体
use super::{Annex, Date, ITagProps, Priorities, Priority, Status, User};
// 引入rocket框架的序列化和反序列化特性
use rocket::serde::{Deserialize, Serialize};

// 定义Todo结构体并为其字段实现序列化和反序列化特性
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(crate = "rocket::serde")]
pub struct Todo {
    // 任务拥有者
    pub owner: String,
    // 任务名称
    pub name: String,
    // 任务优先级
    pub priority: Priorities,
    /// 审核人列表
    pub reviewers: Vec<String>,
    // 执行者列表
    pub performers: Vec<String>,
    // 任务日期
    pub date: Date,
    // 任务标签列表
    pub tags: Vec<ITagProps>,
    // 任务状态
    pub status: Status,
    // 任务描述（可选）
    pub description: Option<String>,
    // 附加信息（可选）
    pub information: Option<String>,
    /// 附件列表（可选）
    pub annexs: Option<Vec<Annex>>,
    // 标记任务是否为焦点任务
    #[serde(rename(serialize = "isFocus"))]
    #[serde(rename(deserialize = "isFocus"))]
    pub is_focus: bool,
}

impl Default for Todo {
    fn default() -> Self {
        Self {
            name: Default::default(),
            priority: Default::default(),
            reviewers: Default::default(),
            performers: Default::default(),
            date: Default::default(),
            tags: Default::default(),
            status: Default::default(),
            description: None,
            information: None,
            annexs: None,
            is_focus: false,
            owner: "".to_string(),
        }
    }
}

impl Todo {
    // 设置任务拥有者
    pub fn set_owner(&mut self, id: String) {
        self.owner = id;
    }

    // 检查是否有审核人
    pub fn have_reviewers(&self) -> bool {
        !self.reviewers.is_empty()
    }

    // 检查是否有执行者
    pub fn have_performers(&self) -> bool {
        !self.performers.is_empty()
    }

    // 判断任务是否为个人任务（无审核人和执行者）
    pub fn is_self_todo(&self) -> bool {
        !self.have_reviewers() && !self.have_performers()
    }

    // 判断任务是否为团队任务
    pub fn is_team_todo(&self) -> bool {
        !self.is_self_todo()
    }

    // 获取任务优先级
    pub fn priority(&self) -> Priorities {
        self.priority.clone()
    }

    // 从给定的值创建一个Todo实例
    pub fn from(value: dto::Todo, owner: &str) -> (String, Self) {
        let id = value.id().to_string();
        let reviewers = value
            .reviewers
            .into_iter()
            .map(|x| x.username().to_string())
            .collect::<Vec<String>>();
        let performers = value
            .performers
            .into_iter()
            .map(|x| x.username().to_string())
            .collect::<Vec<String>>();

        let todo = Todo {
            owner: owner.to_string(),
            name: value.name,
            priority: value.priority,
            reviewers,
            performers,
            date: value.date,
            tags: value.tags,
            status: value.status,
            description: value.description,
            information: value.information,
            annexs: value.annexs,
            is_focus: value.is_focus,
        };

        (id, todo)
    }
}

// 定义一个结构体用于管理不同优先级的任务ID集合
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(crate = "rocket::serde")]
pub struct TodoBox {
    // 低优先级任务ID集合
    pub low: Vec<String>,
    // 中优先级任务ID集合
    pub mid: Vec<String>,
    // 高优先级任务ID集合
    pub fatal: Vec<String>,
    // 关注的任务ID集合
    pub focus: Vec<String>,
    // 历史任务ID集合
    pub history: Vec<String>,
}

impl Default for TodoBox {
    fn default() -> Self {
        Self {
            low: Default::default(),
            mid: Default::default(),
            fatal: Default::default(),
            focus: Default::default(),
            history: Default::default(),
        }
    }
}

impl TodoBox {
    // 从各优先级集合中移除指定的任务ID
    pub fn remove(&mut self, id: &str) {
        let TodoBox {
            low,
            mid,
            fatal,
            focus,
            history,
        } = self;
        // 对low, mid, fatal, focus 进行过滤
        self.low = low.clone().into_iter().filter(|x| x.ne(id)).collect();
        self.mid = mid.clone().into_iter().filter(|x| x.ne(id)).collect();
        self.fatal = fatal.clone().into_iter().filter(|x| x.ne(id)).collect();
        self.focus = focus.clone().into_iter().filter(|x| x.ne(id)).collect();
    }
}
